src/envs/coffee.py

```diff
@@ -0,0 +1,756 @@
+
+    def _Pour_policy(self, state: State, memory: Dict,
+                     objects: Sequence[Object], params: Array) -> Action:
+        # This policy moves the robot next to the cup and then pours until
+        # the cup is filled. Note that if starting out at the end of another
+        # pour, we need to start by rotating the cup to prevent any further
+        # pouring until we've moved over the next cup.
+        del memory, params  # unused
+        move_tilt = self.tilt_lb
+        pour_tilt = self.tilt_ub
+        robot, jug, cup = objects
+        robot_x = state.get(robot, "x")
+        robot_y = state.get(robot, "y")
+        robot_z = state.get(robot, "z")
+        robot_pos = (robot_x, robot_y, robot_z)
+        tilt = state.get(robot, "tilt")
+        jug_x = state.get(jug, "x")
+        jug_y = state.get(jug, "y")
+        jug_z = self._get_jug_z(state, jug)
+        jug_pos = (jug_x, jug_y, jug_z)
+        pour_x, pour_y, _ = pour_pos = self._get_pour_position(state, cup)
+        # If we're close enough to the pour position, pour.
+        sq_dist_to_pour = np.sum(np.subtract(jug_pos, pour_pos)**2)
+        if sq_dist_to_pour < self.pour_pos_tol:
+            dtilt = pour_tilt - tilt
+            return self._get_move_action(jug_pos, jug_pos, dtilt=dtilt)
+        dtilt = move_tilt - tilt
+        # If we're above the pour position, move down to pour.
+        xy_pour_sq_dist = (jug_x - pour_x)**2 + (jug_y - pour_y)**2
+        if xy_pour_sq_dist < self.safe_z_tol:
+            return self._get_move_action(pour_pos, jug_pos, dtilt=dtilt)
+        # If we're at a safe height, move toward above the pour position.
+        if (robot_z - self.robot_init_z)**2 < self.safe_z_tol:
+            return self._get_move_action((pour_x, pour_y, jug_z),
+                                         jug_pos,
+                                         dtilt=dtilt)
+        # Move to a safe moving height.
+        return self._get_move_action((robot_x, robot_y, self.robot_init_z),
+                                     robot_pos,
+                                     dtilt=dtilt)
+
+    def _Pour_terminal(self, state: State, memory: Dict,
+                       objects: Sequence[Object], params: Array) -> bool:
+        del memory, params  # unused
+        _, _, cup = objects
+        return self._CupFilled_holds(state, [cup])
+
+    def _get_jug_handle_grasp(self, state: State,
+                              jug: Object) -> Tuple[float, float, float]:
+        target_x = state.get(jug, "x") + self.jug_handle_x_offset
+        target_y = state.get(jug, "y") + self.jug_handle_y_offset
+        target_z = self.jug_handle_height
+        return (target_x, target_y, target_z)
+
+    def _get_jug_z(self, state: State, jug: Object) -> float:
+        if state.get(jug, "is_held") > 0.5:
+            # Offset to account for handle.
+            return state.get(self._robot, "z") - self.jug_handle_height
```

Comment:
i thought the jug's handle is in the y direction not the z direction

Reply:
both!